### PR TITLE
Negative LIMIT specification

### PR DESCRIPTION
**Pull request by Joeywzr**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added Negative LIMIT specification. And it can be combined with negative offset (but not with positive offset).

Detailed description / Documentation draft:
#28913

### CODE REVIEW

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -824,13 +843,14 @@ static UInt64 getLimitUIntValue(const ASTPtr & node, ContextPtr context, const s
...
+    Field converted = convertFieldToType(field, DataTypeInt128());
     if (converted.isNull())
         throw Exception(
-            "The value " + applyVisitor(FieldVisitorToString(), field) + " of " + expr + " expression is not representable as UInt64",
+            "The value " + applyVisitor(FieldVisitorToString(), field) + " of " + expr + " expression is not representable as Int128",
             ErrorCodes::INVALID_LIMIT_EXPRESSION);
+    Int128 res = converted.safeGet<Int128>();
 
-    return converted.safeGet<UInt64>();
+    return res >= 0 ? UInt64(res) : UInt64(-res);
```
Comment (Avogar) [inline src/Interpreters/InterpreterSelectQuery.cpp:853]: Converting `Int128 -> UInt64` can lead to some special effects:
```sql
:) select * from system.numbers LIMIT 18446744073709552616

SELECT *
FROM system.numbers
LIMIT 18446744073709552000.

Query id: f2d33e9d-dcbe-4464-92fc-171144d8817a

Ok.

0 rows in set. Elapsed: 0.008 sec.
```
We should check that `|res| < UINT64_MAX`

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 2]
```diff
@@ -843,7 +863,13 @@ static std::pair<UInt64, UInt64> getLimitLengthAndOffset(const ASTSelectQuery &
     {
         length = getLimitUIntValue(query.limitLength(), context, "LIMIT");
         if (query.limitOffset() && length)
+        {
             offset = getLimitUIntValue(query.limitOffset(), context, "OFFSET");
+            if (isLimitOrOffsetPositive(query.limitLength(), context, "LIMIT") ^ isLimitOrOffsetPositive(query.limitOffset(), context, "OFFSET"))
+                throw Exception(
+                    "The values of LIMIT and OFFSET should be positive or negative at the same time",
+                    ErrorCodes::INVALID_OFFSET_EXPRESSION);
```
Comment (Avogar) [inline src/Interpreters/InterpreterSelectQuery.cpp:871]: Maybe we can also support `LIMIT -N OFFSET 0`?

File: src/Processors/LimitTransform.cpp [hunk 3]
```diff
@@ -257,89 +256,161 @@ void LimitTransform::splitChunk(PortsData & data)
...
-    /// check if other rows in current block equals to last one in limit
-    if (with_ties && length)
+            length = current_row_num - start;
+        }
+    }
+    else
     {
-        UInt64 current_row_num = start + length;
-        previous_row_chunk = makeChunkWithPreviousRow(data.current_chunk, current_row_num - 1);
+        limit_is_unreachable = (limit > std::numeric_limits<UInt64>::max() - offset);
```
Comment (Avogar) [inline src/Processors/LimitTransform.cpp:352]: This line is exactly the same as line 265, we can move it before if/else statement I guess

**Comment by alexey-milovidov** [comment]
@Joeywzr Could you please update this PR?

**Comment by Joeywzr** [comment]
@alexey-milovidov Sorry, I'm a little busy recently. I didn't have time to finish the PR. I will try to finish it in January.

**Comment by alexey-milovidov** [comment]
No problem, January is ok.

File: src/Processors/LimitTransform.cpp [hunk 4]
```diff
@@ -226,27 +229,106 @@ LimitTransform::Status LimitTransform::preparePair(PortsData & data)
...
+        }
+        else
+            /// This function may be heavy to execute in prepare. But it happens no more than twice, and make code simpler.
+            splitChunk(data);
     }
     else
-        /// This function may be heavy to execute in prepare. But it happens no more than twice, and make code simpler.
-        splitChunk(data);
+    {
+        if (rows == 8192 || rows == 16384 || rows == 32768 || rows == 65536 || rows == 65505)
```
Comment (Joeywzr) [inline src/Processors/LimitTransform.cpp:249]: I want to know if there is a way to locate the last block. Now I can only use this clumsy method to identify, which is not accurate.

Comment (Joeywzr) [reply src/Processors/LimitTransform.cpp:249]: @Avogar @alexey-milovidov

**Comment by Joeywzr** [comment]
please add test label @Avogar @alexey-milovidov

**Comment by alexey-milovidov** [comment]
@Joeywzr let's continue.

**Comment by alexey-milovidov** [comment]
Continued in #42447.

**Pull request closed without merging**